Source/Parallelization/GuardCellManager.cpp

```diff
@@ -184,11 +190,37 @@ guardCellManager::Init (
         ng_FieldSolver = ng_alloc_EB;
         ng_FieldSolverF = ng_alloc_EB;
         ng_FieldSolverG = ng_alloc_EB;
-    } else {
-        ng_FieldSolver = IntVect(AMREX_D_DECL(1, 1, 1));
-        ng_FieldSolverF = IntVect(AMREX_D_DECL(1, 1, 1));
-        ng_FieldSolverG = IntVect(AMREX_D_DECL(1, 1, 1));
     }
+#ifdef WARPX_DIM_RZ
+    else if (maxwell_solver_id == MaxwellSolverAlgo::Yee) {
+        ng_FieldSolver  = CylindricalYeeAlgorithm::GetMaxGuardCell();
+        ng_FieldSolverF = CylindricalYeeAlgorithm::GetMaxGuardCell();
+        ng_FieldSolverG = CylindricalYeeAlgorithm::GetMaxGuardCell();
+    }
+#else
+    else {
+        if (do_nodal) {
+            ng_FieldSolver  = CartesianNodalAlgorithm::GetMaxGuardCell();
+            ng_FieldSolverF = CartesianNodalAlgorithm::GetMaxGuardCell();
+            ng_FieldSolverG = CartesianNodalAlgorithm::GetMaxGuardCell();
+        } else if (maxwell_solver_id == MaxwellSolverAlgo::Yee) {
+            ng_FieldSolver  = CartesianYeeAlgorithm::GetMaxGuardCell();
+            ng_FieldSolverF = CartesianYeeAlgorithm::GetMaxGuardCell();
+            ng_FieldSolverG = CartesianYeeAlgorithm::GetMaxGuardCell();
+        } else if (maxwell_solver_id == MaxwellSolverAlgo::CKC) {
+            ng_FieldSolver  = CartesianCKCAlgorithm::GetMaxGuardCell();
+            ng_FieldSolverF = CartesianCKCAlgorithm::GetMaxGuardCell();
+            ng_FieldSolverG = CartesianCKCAlgorithm::GetMaxGuardCell();
+        }
+    }
+#endif
+
+    // Number of guard cells is the max of that determined by particle shape factor and
+    // the stencil used in the field solve
+    ng_alloc_EB.max( ng_FieldSolver );
+    ng_alloc_F.max( ng_FieldSolverF );
+    ng_alloc_G.max( ng_FieldSolverG );
+    ng_alloc_J.max( ng_FieldSolver  );
```

Comment:
@RemiLehe I wanted to bring this to your attend that I added this last line for ng_alloc_J since current-density is used in EM-push and it may be required for test-cases without particles.

Reply:
Yes, it seems that anyways we want all MultiFabs to have the same number of guard cells (the overall maximum):
https://github.com/ECP-WarpX/WarpX/blob/7a1a10f5c711d064ec8a5a1caa9fdecd36bad162/Source/Parallelization/GuardCellManager.cpp#L163-L178
It's true that the charge and current density are supposed to be zero without particles, but I think the actual MultiFabs are still used in the field solver even in that case (they are just filled with zeros).
So I think it makes sense to add `ng_alloc_J.max(ng_FieldSolver)` as you did.
Should we add `ng_alloc_Rho`, too (since the charge density is also used in the field solver)?

@RemiLehe 
Does what Reva and I are saying here make sense to you?

Reply:
ah yes ! indeed!

Reply:
Hm, actually I think it should be fine if `J` does not have as many guard cells as `E` and `B`.
The reason is that the Maxwell push typically looks like:
```
Ex(i,j,k) += dt/eps0 * Jx(i,j,k) + dt/dy * ( Bz(i,j+1,k) - Bz(i,j,k) )  // plus another similar term for the z derivative
```
The above equation is only applied inside the *valid cells* of each sub-domain. But because of `j+1`, we need to know `B` in the guard cells (i.e. we need at least one guard cell). However, because `Jx` is taken at the same `i,j,k` index as `Ex`, I don't think that we need guard cells for `Jx`.
Does this make sense?

Reply:
ah I see! Sounds good. so this is not needed for `J`. Okay I will remove that line. Thank you Remi